src/common/mfu_daos.c

```diff
@@ -677,3 +688,326 @@ int daos_cleanup(
+                    daos_key_t aiov;
+                    daos_iod_t iod;
+		    daos_recx_t recx;
+                    snprintf(akey, akey_kds[j].kd_key_len + 1, "%s", akey_ptr);
+                    d_iov_set(&aiov, (void*)akey, akey_kds[j].kd_key_len);
+
+                    /* set iod values */
+                    iod.iod_nr    = 1;
+                    iod.iod_type  = DAOS_IOD_SINGLE;
+                    iod.iod_size  = DAOS_REC_ANY;
+		    iod.iod_recxs = NULL;
+
+                    d_iov_set(&iod.iod_name, (void*)akey, strlen(akey));
+
+                    /* Do a fetch (with NULL sgl) of single value type, and if that
+		     * returns iod_size == 0, then a single value does not exist. */
+                    rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, &diov, 1, &iod, NULL, NULL, NULL);
+                    if (rc != 0) {
+                        MFU_LOG(MFU_LOG_ERR, "DAOS daos_obj_fetch returned with errors: ", MFU_ERRF,
+	                        MFU_ERRP(-MFU_ERR_DAOS));
+                    }
+
+                    /* if iod_size == 0 then this is a DAOS_IOD_ARRAY type */
+                    if ((int)iod.iod_size == 0) {
+                        rc = daos_copy_recx_array(&diov, &aiov, src_oh, dst_oh, &iod);
+                        if (rc != 0) {
+                            MFU_LOG(MFU_LOG_ERR, "DAOS daos_copy_recx_array returned with errors: ", MFU_ERRF,
+	                            MFU_ERRP(-MFU_ERR_DAOS));
+                        }
+                    } else {
+                        rc = daos_copy_recx_single(&diov, src_oh, dst_oh, &iod);
+                        if (rc != 0) {
+                            MFU_LOG(MFU_LOG_ERR, "DAOS daos_copy_recx_single returned with errors: ", MFU_ERRF,
+	                            MFU_ERRP(-MFU_ERR_DAOS));
+                        }
+                    }
+
+                    /* advance to next akey returned */	
+                    akey_ptr += akey_kds[j].kd_key_len;
+                }
+            }
+        }
+    }
+    return rc;
+}
+
+int daos_obj_copy(daos_args_t* da,
+                          flist_t* flist) {
+    int rc;
+    uint64_t i;
+    const elem_t* p = flist->list_head;
+
+    for (i = 0; i < flist->list_count; i++) {
+        /* open DAOS object based on oid[i] to get obj handle */
+        daos_handle_t oh;
+        daos_obj_id_t oid;
+        oid.lo = p->obj_id_lo;	
+        oid.hi = p->obj_id_hi;	
+        rc = daos_obj_open(da->src_coh, oid, 0, &oh, NULL);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "DAOS object open returned with errors: ", MFU_ERRF,
+	            MFU_ERRP(-MFU_ERR_DAOS));
+	    return 1;
+        }
+
+        /* open handle of object in dst container */
+        daos_handle_t dst_oh;
+        rc = daos_obj_open(da->dst_coh, oid, 0, &dst_oh, NULL);
+        if (rc != 0) {
+            MFU_LOG(MFU_LOG_ERR, "DAOS object open returned with errors: ", MFU_ERRF,
+	            MFU_ERRP(-MFU_ERR_DAOS));
+	    return 1;
+        }
```

Comment:
For an error on opening the destination, we need to make sure we close the source:
`daos_obj_close(oh, NULL);`